### PR TITLE
Add more negative stride support to BLAS Level 1/2 functions

### PR DESCRIPTION
**Pull request by N5N3**
This PR could be regarded as the follow-up attempt of JuliaLang/julia#41513 and JuliaLang/julia#39751

In general, negative stride support is added to the following Level 1 functions:
(Let `st` denotes the input vector's stride)
1. `nrm2`, `asum`
Openblas always return `0` if `st < 0`, so I pass in `abs(st)`.
2. `scal!`
`st` and `-st` give the same result, so I pass in `abs(st)`.
3. `axpy(by)!`, `BLAS.copyto!`
`st` does affect the results' order so I have to pass negative `st` directly. At least OpenBLAS and MKL could handle these calls.
(API with stide input, like `rot!(n, X, incx, Y, incy, c, s)` is not changed in this PR. User should pass in correct pointer if `inc < 0`)

Besides, `st < 0` is supported by MKL and openblas for all the implemented Level 2 functions, so all of them are extended in this PR.

Test Added.

Edit: `spr!` added

Edit2: commits have been cleaned for easier review.

### CODE REVIEW

**Comment by N5N3** [comment]
`iamax` with `st < 0` failed on `linuxaarch64`.
Based on MKL's document:
["If either `n` or `incx` are not positive, the routine returns 0."](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-fortran/top/blas-and-sparse-blas-routines/blas-routines/blas-level-1-routines-and-functions/i-amax.html?wapkw=iamax?wapkw=iamax) 
and OpenBLAS x86-64 source code:
https://github.com/xianyi/OpenBLAS/blob/2d7ca63e21e97dcd2db5c3aa7658f6cc2a504e55/kernel/x86_64/iamax.S#L72-L75
`0` shoud be the expected return for negetive `st`.

Looks like openblas on arm only test `n < 0`: https://github.com/xianyi/OpenBLAS/blob/2d7ca63e21e97dcd2db5c3aa7658f6cc2a504e55/kernel/arm64/iamax_thunderx2t99.c#L215
Should we threat this inconsistency as "bug"?

Anyway, `iamax` should be seldomly used so I'll add a negetive stride check in Julia side.

File: stdlib/LinearAlgebra/src/blas.jl [hunk 1]
```diff
@@ -159,6 +159,18 @@ end
 
 
 # Level 1
+isdense(x) = x isa DenseArray
+isdense(x::Base.FastContiguousSubArray) = isdense(parent(x))
+isdense(x::Base.ReshapedArray) = isdense(parent(x))
+isdense(x::Base.ReinterpretArray) = isdense(parent(x))
+@inline function ptrst1(x::AbstractArray)
```
[the remark this answers is not in the document]
Comment (N5N3) [reply stdlib/LinearAlgebra/src/blas.jl:166]: Make sense. Changed

File: stdlib/LinearAlgebra/src/blas.jl [hunk 2]
```diff
@@ -159,6 +159,18 @@ end
...
+    isdense(x) && return pointer(x), 1 # simpify runtime check when possibe
+    ndims(x) == 1 || strides(x) == Base.size_to_strides(strides(x, 1), size(x)...) ||
```
[the remark this answers is not in the document]
Comment (N5N3) [reply stdlib/LinearAlgebra/src/blas.jl]: Nice catch. fixed. (Test added)

File: stdlib/LinearAlgebra/src/blas.jl [hunk 3]
```diff
@@ -159,6 +159,18 @@ end
...
+        throw(ArgumentError("only support vector like inputs"))
+    st = stride(x, 1)
+    ptr = st > 0 ? pointer(x) : pointer(x, lastindex(x))
```
[the remark this answers is not in the document]
Comment (N5N3) [reply stdlib/LinearAlgebra/src/blas.jl:171]: Based on my limited test, MKL can handle `st == 0` correctly if the result is meaningful.

File: stdlib/LinearAlgebra/src/blas.jl [hunk 4]
```diff
@@ -249,7 +261,10 @@ for (fname, elty) in ((:dscal_,:Float64),
             DX
         end
 
-        scal!(DA::$elty, DX::AbstractArray{$elty}) = scal!(length(DX),DA,DX,stride(DX,1))
+        scal!(DA::$elty, DX::AbstractArray{$elty}) = let (p, st) = ptrst1(DX)
```
[the remark this answers is not in the document]
Comment (N5N3) [reply stdlib/LinearAlgebra/src/blas.jl]: `let` has been removed.

File: stdlib/LinearAlgebra/test/blas.jl [hunk 5]
```diff
@@ -4,20 +4,31 @@ module TestBLAS
...
+    for j in 1:n, i in (uplo==:L ? (j:n) : (1:j))
+        push!(AP, A[i,j])
+    end
+    return AP
+end
 
+@testset "vec_pointer_stride" begin
+    a = zeros(4,4,4)
+    @test BLAS.asum(view(a,1:2:4,:,:)) == 0 # vector like
+    @test_throws ArgumentError BLAS.asum(view(a,1:3:4,:,:)) == 0 # non-vector like
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/test/blas.jl:23]:
```suggestion
    @test_throws ArgumentError BLAS.asum(view(a,1:3:4,:,:)) # non-vector like
```

**Review by dkarrasch — APPROVED** [review]
This looks good to me, just one very minor comment.

File: stdlib/LinearAlgebra/src/blas.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/blas.jl:176]: Do we have at least one "meaningful" test for this case?

**Comment by dkarrasch** [comment]
BTW, does this also fix https://github.com/JuliaLang/julia/pull/41513#issuecomment-905238030? Does it therefore "replace" #42054 on master, so #42054 should be only backported to v1.7 and v1.6, but should not be used on master?

File: stdlib/LinearAlgebra/src/blas.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (N5N3) [reply stdlib/LinearAlgebra/src/blas.jl:176]: No, I'll add some.

**Comment by N5N3** [comment]
> BTW, does this also fix https://github.com/JuliaLang/julia/pull/41513#issuecomment-905238030? Does it therefore "replace" https://github.com/JuliaLang/julia/pull/42054 on master, so https://github.com/JuliaLang/julia/pull/42054 should be only backported to v1.7 and v1.6, but should not be used on master?
No, that comment is not fixed in this PR.

**Comment by dkarrasch** [comment]
@N5N3 When you find the time, could you please resolve the merge conflict (on a quick glance, I couldn't find `pA`, the pointer to `A` defined in your version?)? Once tests pass, we can merge.

**Comment by N5N3** [comment]
The unreviewed changes are sepetated.
Note: `LinearAlgebra.gemv!` won't call BLAS if it found a 0-stride src vector. 
I'm not sure whether OpenBLAS's `axpy`/`dot` work well with 0-stride input on all platform. Let's check it with CI.

File: stdlib/LinearAlgebra/test/matmul.jl [hunk 6]
```diff
@@ -297,6 +297,15 @@ end
     end
 end
 
+@testset "matrix x vector with negtive lda or 0 stride" for T in (Float32, Float64)
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/test/matmul.jl:300]:
```suggestion
@testset "matrix x vector with negative lda or 0 stride" for T in (Float32, Float64)
```

**Comment by N5N3** [comment]
Error on linux64, win32, win64 should be unrelated. (BLAS test passed.)
It is good that linuxaarch64 is green.
But freebsd64 seems hang?

**Comment by N5N3** [comment]
@[dkarrasch](https://github.com/dkarrasch) All tests are done. Error on linux32/64 seems unrelated. (BLAS test passed).
I guess it OK to merge?

**Pull request merged**